File: packages/protocol/src/tasks.ts
/**
 * The task list a session carries.
 *
 * One tool replaces the whole list on each call, so there is no patch format and
 * no per-task id: the model sends what the list should now be, and the last call
 * wins. That is the whole of the write contract, and it is why a task is two
 * fields.
 *
 * It lives here rather than beside the tool for the reason `subagent.ts` does:
 * six layers read it and none of them can reach the others. The tool writes it,
 * the store holds it, the prompt renders it, the terminal draws it, a chat
 * command prints it and the browser fetches it over REST.
 *
 * The list is in the session's metadata bag rather than a column, the way
 * lineage is. It costs no schema, no index and no query surface, and nothing
 * searches by it. A subagent opens its own session, so keying by session is what
 * gives a delegated run its own list for free.
 *
 * The browser only ever *reads* a list, over REST, so this is the shapes and
 * nothing else — there is no `tasksOf` here to mirror the server's, because
 * nothing in the browser holds a raw metadata bag.
 */

import { z } from 'zod';

/** Where a session records its task list. */
export const TASKS_METADATA_KEY = 'tasks';

/**
 * How many tasks a list may hold.
 *
 * Ten, because the list is re-read on every iteration of every turn and a plan
 * nobody can hold in their head is not a plan. Work that needs more of them
 * needs a smaller first task.
 */
export const MAX_TASKS = 10;

/** How long one task's text may be. */
export const MAX_TASK_CHARS = 100;

/** Where one task has got to. At most one task may be `doing`. */
export const TaskStatusSchema = z.enum(['todo', 'doing', 'done']);
export type TaskStatus = z.infer<typeof TaskStatusSchema>;

/** One line of the plan. */
export const TaskItemSchema = z.object({
  text: z.string().min(1).max(MAX_TASK_CHARS),
  status: TaskStatusSchema,
});
export type TaskItem = z.infer<typeof TaskItemSchema>;

File: crates/tools/src/tasks.rs
//! Where a session's task list actually gets written.
//!
//! The store lives in `darkwire-core`, and the session key the list belongs to
//! is not something a tool may name: a model writing its own plan must not be
//! able to write it onto somebody else's conversation. So this follows the
//! answer [`automation`](crate::automation) gives — **the interface is declared
//! down here, and the composition root supplies the implementation** — with the
//! same guard on the same reasoning: the port is bound to the turn that got it,
//! and closes over the session key rather than taking one.
//!
//! That binding is also what gives a subagent its own list for nothing. A
//! delegated run opens its own session, so the port its turn receives points
//! somewhere else, and neither run can see or clear the other's plan.
//!
//! One method, because the write contract is one call: the model sends the list
//! it wants, and the last call wins. There is no add, no complete and no
//! reorder, so there is nothing here to keep consistent with a patch format.

use darkwire_core::Result;
use darkwire_protocol::tasks::TaskItem;

/// One turn's access to its session's task list, already scoped to it.
pub trait TaskPort: Send + Sync {
    /// Replaces the whole list. An empty slice clears it.
    fn replace(&self, tasks: &[TaskItem]) -> Result<()>;
}
